### ISSUE TITLE
missing rule: enforce multiple var declarations [$30]

### ISSUE THREAD

**Issue by remitbri** [eslint/eslint#1619]
Hi. I've tried to review each and every rule of JSHint, JSCS and ESLint to see if I could entirely "get rid of" the first two in favour of the latter one. As suggested by the owner of the @geteslint twitter account, I'm reporting the JSCS rules I've found actually not covered by ESLint. My apologies if it's a duplicate.

 - disallowMultipleVarDecl

I know there's the `one-var` rule, but what about enforcing the contrary? I.e enforcing a `var` for each declaration.

Actually JSCS has three settings :
 - `true` disallows multiple variable declarations except within a for loop
 - 'strict' disallows all multiple variable declarations
 - 'exceptUndefined' allows declarations where all variables are not defined, e.g

 ```js
 var a, b; // valid with 'exceptUndefined'
 var x = 1;  // valid with 'true', 'strict' and 'exceptUndefined'
 var y = 2;  // valid with 'true', 'strict' and 'exceptUndefined'
 for (var i = 0, j = arr.length; i < j; i++) {}  // valid with `true' and `exceptUndefined'

 var x = 1, // invalid
     y = 2; // invalid
 var x, y = 2, z; // invalid
 ```

**Comment by xjamundx** [comment]
https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/disallow-multiple-var-decl.js

**Comment by Flet** [comment]
Is this rule really missing? In jscs `disallow-multiple-var-decl` maps to JSHint's `onevar`, which is implemented in run `one-var` in eslint http://eslint.org/docs/rules/one-var.html

**Comment by nzakas** [comment]
I believe he is requesting the opposite of that rule. @remitbri is that correct?

**Comment by Flet** [comment]
Ah my mistake. JSCS references `onevar` in its documentation, but reading the jslint docs, its clear it does not map. :+1:

**Comment by remitbri** [comment]
@nzakas Correct!

**Comment by feross** [comment]
I just put up a $15 bounty on this issue.

See: https://www.bountysource.com/issues/7465930-missing-rule-enforce-multiple-var-declarations

**Comment by feross** [comment]
Aaannd, @othiym23 just matched that, bringing the total bounty to $30.

https://www.bountysource.com/issues/7465930-missing-rule-enforce-multiple-var-declarations

**Comment by nzakas** [comment]
For anyone who wants to try: this would basically change `one-var` to accept `always` (default) or `never` as options.

**Comment by xjamundx** [comment]
yay I want this `never`! May give it a stab tomorrow

**Comment by feross** [comment]
Would be nice to support the option to allow multiple declarations when all variables are undefined. 

This would be valid:

```js
var a, b, c;
var d = 42;
```

Invalid:

```js
var a, b, c = 2;
var d = 42;
```

**Comment by feross** [comment]
Nice, @xjamundx!

**Comment by xjamundx** [comment]
beaten to the punch!

**Comment by Flet** [comment]
@dannyfritz you are fast :)

**Comment by dannyfritz** [comment]
I don't know how much i like the `always` and `never`. I agree with the behavior, but it seems like `never` doesn't really allude to allowing a single declaration. Also, my PR doesn't do the all `undefined` var statement.

**Comment by xjamundx** [comment]
What did people think about allowing for multiple `var` in the case of pure initialization ` var x,y,z`? I believe most people that want `never`, actually want to combine multiple initializations into a single declaration. Here's a popular version of that:
http://benalman.com/news/2012/05/multiple-var-statements-javascript/#highlighter_996633

Happy if that's a completely separate rule though.

**Comment by Flet** [comment]
@dannyfritz here is my attempt at `"only-undefined"` on top of your PR:
https://github.com/Flet/eslint/commit/5febbde5a3fefc95634645b76b3c31b351f60e04

**Comment by xjamundx** [comment]
Presumably these won't be affected by this rule:

```js
var {x, y, z} = a;
var [a, ...b] = c;
```

Maybe they deserve a test? Any other ES6 things that could cause some sort of conflict?

```js
import {x, y} from 'z'; // should be fine
let a b, c; // should be covered
class X {} // not relevant
```

Probably good, just double check to make sure `never` doesn't accidently flag destructuring?

**Comment by dannyfritz** [comment]
@xjamundx I wrote a test for
```js
function foo() { var a = [1, 2, 3]; var [b, c, d] = a; }
```
but it fails on `=`. I'm assuming it doesn't support destructuring?
```
  1 failing

  1) one-var function foo() { var a = [1, 2, 3]; var [b, c, d] = a; }:
     AssertionError: Should have no errors but had 1: [ { fatal: true,
    severity: 2,
    message: 'Unexpected token =',
    line: 1,
    column: 51 } ]: expected 1 to equal 0
```

**Comment by xjamundx** [comment]
you have to add `{ecmaFeatures: { destructuring: true }` into the config
for that rule!

**Comment by dannyfritz** [comment]
@xjamundx Indeed. :) Updated to include the tests, doesn't seem to cause any issues.

**Comment by dannyfritz** [comment]
@xjamundx for the var's on top setting, I wrote a rule already to do that. :wink: http://eslint.org/docs/rules/vars-on-top.html

**Comment by nzakas** [comment]
I really don't like `only-undefined`. That's also a completely separate feature request that should be discussed on a separate issue. Let's keep the discussion here to always vs. never.

**Comment by Flet** [comment]
@nzakas Roger that!
